### ISSUE TITLE
Add support for travelling sales routing, Z-level and turn restrictions

### ISSUE THREAD

**Issue by paulwittle** [qgis/QGIS#32295]
**Feature description.**
The default routing tools in the processing toolbox can use a local road network and account for one-way streets however I can't see any settings which would allow the tools to take account of turn restrictions (banned turns) or Z-level attributes.

Without these key pieces of information the tools are unable to provide accurate routing as they are vital to correct calculation. To give a UK specific example; if you are using Ordnance Survey's Highways network then some bridge links are split with the upper level links have a Z value of 1 whilst the lower links have a value of 0. The default tools permit routing to effectively 'jump off' the bridge onto the lower section and vice versa.

It is much the same for the banned turns although creating a fix for these may be technically more complex.

It would also be useful to look at whether or not travelling sales person routing could be added as default. To date I have written my own processing scripts which allow routing with way-points so perhaps that might be a starting point.

I would like to get involved in trying to add support for these functions as I think they will provide a significant boost to the default functionality and provide a good level of support before requiring third party options such as paid for routing software or the use of API keys (which may also incur a cost).

**Proposed actions.**
Given that I am new to the group and need to get up to speed on development principals / scripting languages; I would like to propose that I organise a coding sprint or session of some sort where we can discuss / develop options and in effect kick start development.

I will continue to develop some options using my own processing scripts and they can be used as a basis for discussion however I think a more collaborative approach may be required to find a good quality solution.

For turn restrictions I'm wondering whether that is best done by modifying the routing tree calculations (probably the right method but complex) or by the development of a tool which modifies the network itself.

The later option is probably wrong on every level but it might be fairly simple to achieve by adding fake links and perhaps overriding the link distance / speed settings. I will work on explaining that better prior to any code sprint if that is thought to be a good route forward.

**Further points.**
Would also like to investigate the speed issues; my processing scripts currently clip the network prior to running routing as otherwise it takes over an hour for any process to run. This is because the routing tree calculations appear to work out the whole network rather than doing any form of base iteration to work out how much of the network is required.

Additionally I would like to investigate why the routing functions are so hard to stop. We have had general issues where QGIS starts a task and it is difficult or perhaps even impossible to stop the tasks from QGIS or perhaps even via the Windows task manager. The organisation I work for think this should be resolved as part of this work if possible.

It has been noted that pgRouting provides the functionality described so for additional clarification the company I work for is not currently in a position to migrate to PostGIS (although we do plan to eventually). I feel there may be benefit in improving the default tools to cover some additional functionality even if it does only serve as a stepping stone to using larger routing applications. Many of those application, pgRouting included, may be too complex for some users but they may still want to be able to achieve basic network analysis in support of planning work for example.

**Comment by nyalldawson** [comment]
> Would also like to investigate the speed issues; my processing scripts currently clip the network prior to running routing as otherwise it takes over an hour for any process to run. This is because the routing tree calculations appear to work out the whole network rather than doing any form of base iteration to work out how much of the network is required.

This is basically because the current QGIS implementation use the inferior Dijkstra algorithm, and needs updating to use the a* algorithm instead.

> Additionally I would like to investigate why the routing functions are so hard to stop. 

This should be a relatively trivial fix -- it would involve adding new `QgsFeedback* feedback = nullptr` arguments to the various graph building and routing functions, and then periodically checking the feedback to detect whether the building/routing should abort out early, e.g.:

 ```
if ( feedback && feedback->isCanceled() )
{
   // cleanup and abort gracefully
   ...
   return;
}
```

Looks like QgsVectorLayerDirector::makeGraph already does this in some places, but if that's where the slow cancelation still occurs then similar checks might need adding in more of the loops in this function. Other parts of the routing process, like QgsGraphAnalyzer::dijkstra and QgsGraphAnalyzer::shortestTree, don't have any early cancelation support yet.

> It has been noted that pgRouting provides the functionality described so for additional clarification the company I work for is not currently in a position to migrate to PostGIS (although we do plan to eventually). I feel there may be benefit in improving the default tools to cover some additional functionality even if it does only serve as a stepping stone to using larger routing applications. Many of those application, pgRouting included, may be too complex for some users but they may still want to be able to achieve basic network analysis in support of planning work for example.

Agreed - there's clearly a demonstrated need for simple, out-of-the-box routing algorithms which don't rely on an external tools or web services and which work with simple line layer-based networks.

**Comment by nyalldawson** [comment]
Also should say -- welcome @paulwittle ! Thanks for showing the right attitude to jump into a project like QGIS, and for declaring in advance your plans to "roll up your sleeves" and contribute. It's a great community, and honestly contributing like you plan to here will push your development and spatial analysis skills to a new level quickly. Any time your employer invests in allowing you to work on this will easily pay off in new skills and expertise ten-fold.

**Comment by wonder-sk** [comment]
> This is basically because the current QGIS implementation use the inferior Dijkstra algorithm, and needs updating to use the a* algorithm instead.

It also depends on how big is the network. Just using A* algorithm may not be enough (still too many nodes to visit). But there are solutions for that, e.g. https://en.wikipedia.org/wiki/Contraction_hierarchies

It may be wise to check some existing open source routing libraries first as implementing all this from scratch might be a headache - it's not just the algorithm itself, you also need some pre-processing tool to create efficient graph representation from input data etc.
https://github.com/Project-OSRM/osrm-backend
https://github.com/valhalla/valhalla

The current implementation of routing graph in QGIS is fairly simple - for any more advanced routing that involves turn restrictions things will need quite radical changes. Right now, if I remember correctly, junctions are nodes and road segments are edges in the routing graph. Applying turn restrictions to this representation is difficult and it would require modification of Dijkstra's algorithm (or any other standard routing alg). A better solution is to "switch" the representations: have road segments as nodes (one node for each direction) and edges connecting roads when such turn is allowed. This approach allows use of standard routing algorithms.

Generally I would be in favor of adopting one of the existing open source routing engines as THE engine for QGIS that would be shipped as a part of QGIS than rewriting things from scratch as it may be a very long project :-)

**Comment by elpaso** [comment]
@wonder-sk I agree with you, I've been using OSRM in production for ~7 years now and I'm still a great fan of it, it works very well and it's **fast**! I think they also do have a library that we could use independently from the main application.

**Comment by paulwittle** [comment]
Thanks for all the really constructive comments everyone. I'm definitely going to do my best to get something going and start investigating the options as it is definitely best not to reinvent the wheel here. 

You have all rightly pointed out that there many options and so is sounds like looking at OSRM would be a good starting point. I think my main goal in this would be to ensure we can do simple things well but without being overly restricted on the network we can use. I think @nyalldawson summarised it brilliantly at the end of his first response.

Slight technical point regarding my planned involvement is that my current employer is in the process of restructuring so I'm thinking it will most likely to take me till early next year to get myself sorted and ready to really start working on things. 

Once the restructure is sorted I'm hoping to host a UK south west user group meeting in the first instance. That should hopefully give me the time and opportunity to get a working build of QGIS so I can get started.  

Lets hope I can finally find the time and employer buy in to get myself up and running this time.

**Comment by roya0045** [comment]
I agree with @wonder-sk that QGIS should 'adopt' a standard routing library. At the moment I rely on the online routing mapper plugin as that is the quickest and simplest way to get a good optimal path without hassle.
